File: kickr-server/src/main/java/com/kickr_server/notification/NotificationService.java
package com.kickr_server.notification;

import com.kickr_server.exception.user.UserNotFoundException;
import com.kickr_server.user.User;
import com.kickr_server.user.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class NotificationService {

    private final NotificationRepository notificationRepository;
    private final UserRepository userRepository;

    @Transactional
    public void createNotification(User recipient, User actor, NotificationType type, String message, String targetId) {
        if (recipient.getId().equals(actor.getId())) {
            return; // Don't notify yourself
        }

        Notification notification = Notification.builder()
                .recipient(recipient)
                .actor(actor)
                .type(type)
                .message(message)
                .targetId(targetId)
                .isRead(false)
                .build();
        notificationRepository.save(notification);
    }

    public List<Notification> getNotificationsForUser(UUID userId) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new UserNotFoundException("User not found"));
        return notificationRepository.findByRecipientOrderByCreatedAtDesc(user);
    }

    public long getUnreadCount(UUID userId) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new UserNotFoundException("User not found"));
        return notificationRepository.countByRecipientAndIsReadFalse(user);
    }

    @Transactional
    public void markAsRead(UUID notificationId) {
        notificationRepository.findById(notificationId).ifPresent(n -> {
            n.setRead(true);
            notificationRepository.save(n);
        });
    }

    @Transactional
    public void markAllAsRead(UUID userId) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new UserNotFoundException("User not found"));
        List<Notification> unread = notificationRepository.findByRecipientOrderByCreatedAtDesc(user)
                .stream().filter(n -> !n.isRead()).toList();
        unread.forEach(n -> n.setRead(true));
        notificationRepository.saveAll(unread);
    }
}

File: kickr-client/src/services/notificationService.ts
import api from './axios';

export type NotificationType = 'FOLLOW' | 'NEW_REVIEW';

export interface Notification {
    id: string;
    recipientId: string;
    actorId: string;
    actorName: string;
    type: NotificationType;
    message: string;
    targetId: string;
    isRead: boolean;
    createdAt: string;
}

export const notificationService = {
    getNotifications: async (userId: string): Promise<Notification[]> => {
        const response = await api.get<Notification[]>(`/notifications/user/${userId}`);
        return response.data;
    },

    getUnreadCount: async (userId: string): Promise<number> => {
        const response = await api.get<number>(`/notifications/unread-count/${userId}`);
        return response.data;
    },

    markAsRead: async (id: string): Promise<void> => {
        await api.post(`/notifications/${id}/read`);
    },

    markAllAsRead: async (userId: string): Promise<void> => {
        await api.post(`/notifications/user/${userId}/read-all`);
    }
};
